### PR TITLE
Fix overlapping axis labels on smaller viewports

### PR DESCRIPTION
**Pull request by krkshitij**
## Previous Behavior

X-axis labels start overlapping when the window is zoomed in, or the window size is reduced

## New Behavior

Scrollbar becomes visible when the viewport is smaller than the minimum width of the cartesian chart

- The minimum width is larger than it should be for some charts with string axis and 2:1 spacing
- Test cases to check if the cartesian charts are readable on smaller viewports:
  - Shrink the window and reload the chart when width and height props are not provided
  - Zoom in the window and reload the chart when width and height props are not provided
  - Decrease the width and height props passed to the chart

### CODE REVIEW

**Comment by fabricteam** [comment]
## 📊 Bundle size report

<sub>🤖 This report was generated against <a href='https://github.com/microsoft/fluentui/commit/cf6b94ba6559236150bc1f1814b023607edaeb20'>cf6b94ba6559236150bc1f1814b023607edaeb20</a></sub>
<div id="bundleSizeComment9423"/>

File: packages/react-charting/src/components/CommonComponents/CartesianChart.base.tsx [hunk 1]
```diff
@@ -682,4 +697,55 @@ export class CartesianChartBase extends React.Component<IModifiedCartesianChartP
...
+    }
+
+    let minChartWidth = this.margins.left! + this.margins.right! + labelWidth * (this._tickValues.length - 1);
+
+    if (
+      [ChartTypes.GroupedVerticalBarChart, ChartTypes.VerticalBarChart, ChartTypes.VerticalStackedBarChart].includes(
+        this.props.chartType,
+      )
+    ) {
+      const minDomainMargin = 8;
```
Comment (AtishayMsft) [inline packages/react-charting/src/components/CommonComponents/CartesianChart.base.tsx:745]: Can you explain min domain margin is there only for vertical bar chart?

File: packages/react-charting/src/utilities/utilities.ts [hunk 2]
```diff
@@ -166,18 +166,20 @@ export function createNumericXAxis(xAxisParams: IXAxisParams, chartType: ChartTy
...
-  const formatMillisecond = locale.format('.%L');
-  const formatSecond = locale.format(':%S');
-  const formatMinute = locale.format('%I:%M');
-  const formatHour = locale.format('%I %p');
-  const formatDay = locale.format('%a %d');
-  const formatWeek = locale.format('%b %d');
-  const formatMonth = locale.format('%B');
-  const formatYear = locale.format('%Y');
+function multiFormat(date: Date, locale?: d3TimeFormat.TimeLocaleObject) {
+  const timeFormat = locale ? locale.format : d3TimeFormat.timeFormat;
```
Comment (AtishayMsft) [inline packages/react-charting/src/utilities/utilities.ts:174]: why this change?

File: packages/react-charting/src/utilities/utilities.ts [hunk 3]
```diff
@@ -240,7 +242,11 @@ export function createDateXAxis(
   if (xAxisElement) {
     d3Select(xAxisElement).call(xAxis).selectAll('text').attr('aria-hidden', 'true');
   }
-  return xAxisScale;
+  const tickValues = (tickParams.tickValues ?? xAxisScale.ticks(xAxisCount)).map((val, idx) => {
+    const tickFormat = xAxis.tickFormat();
+    return tickFormat ? tickFormat(val, idx) : multiFormat(val as Date);
+  });
+  return { xScale: xAxisScale, tickValues };
```
Comment (AtishayMsft) [inline packages/react-charting/src/utilities/utilities.ts:249]: Can we get tickValues from xScale directly in cartesian base

File: packages/react-charting/src/components/CommonComponents/CartesianChart.base.tsx (hunk 1 above)
Comment (krkshitij) [reply packages/react-charting/src/components/CommonComponents/CartesianChart.base.tsx:745]: Yes, we had some padding on the left and right side of the chart when we implemented 2:1 spacing in vertical bar chart, vertical stacked bar chart and grouped vertical bar chart

File: packages/react-charting/src/utilities/utilities.ts (hunk 2 above)
Comment (krkshitij) [reply packages/react-charting/src/utilities/utilities.ts:174]: The tick values can have varying widths depending on the format. As per the docs, d3 formats the dates in a similar way when custom locale is not provided. So for the default case we have used this utility with the default locale.

image

File: packages/react-charting/src/components/CommonComponents/CartesianChart.base.tsx [hunk 4]
```diff
@@ -647,7 +662,7 @@ export class CartesianChartBase extends React.Component<IModifiedCartesianChartP
       }
       if (this.props.parentRef || this.chartContainer) {
         const container = this.props.parentRef ? this.props.parentRef : this.chartContainer;
-        const currentContainerWidth = container.getBoundingClientRect().width;
+        const currentContainerWidth = Math.max(container.getBoundingClientRect().width, this._calculateChartMinWidth());
```
Comment (AtishayMsft) [inline packages/react-charting/src/components/CommonComponents/CartesianChart.base.tsx:665]: This call happens in the second render cycle, right?

File: change/@fluentui-react-charting-1f77b865-250d-4cb3-bb5a-93fbdbf94a7c.json [hunk 5]
```diff
@@ -0,0 +1,7 @@
+{
+  "type": "patch",
```
Comment (AtishayMsft) [inline change/@fluentui-react-charting-1f77b865-250d-4cb3-bb5a-93fbdbf94a7c.json:2]: I think we can get past setting the min width and height prop if we don't touch the existing behavior if we make our min width requirement a bit more relaxed than the current max label length X (no of ticks - 1)

File: packages/react-charting/src/components/CommonComponents/CartesianChart.base.tsx (hunk 4 above)
Comment (krkshitij) [reply packages/react-charting/src/components/CommonComponents/CartesianChart.base.tsx:665]: Yes

**Review by AtishayMsft — APPROVED** [review]
Approving the PR. You can add explaining comments and merge.

**Review by AtishayMsft — COMMENTED** [review]
Approving the PR. You can add explanation comments and merge.
